src/ips.rs

```diff
@@ -0,0 +1,505 @@
+    fn update_rating_by_location(
+        &self,
+        selected_node: &Node,
+        nodes: &[Node],
+        ratings: &mut [(IpAddr, usize, f64)],
+    ) {
+        let selected_location = &selected_node
+            .geolocation
+            .as_ref()
+            .map(|geo_info| {
+                Location::new(
+                    geo_info.latitude.unwrap_or_default(), // TODO(asmie): refactor when zg-core will be updated with Location
+                    geo_info.longitude.unwrap_or_default(),
+                )
+            })
+            .unwrap_or(Location::new(0.0, 0.0));
+
+        for (node_idx, node) in nodes.iter().enumerate() {
+            let mut rating = 0.0;
+            if let Some(geo_info) = &node.geolocation {
+                let location = Location::new(
+                    geo_info.latitude.unwrap_or_default(),
+                    geo_info.longitude.unwrap_or_default(),
+                );
+                let distance = selected_location.distance_to(&location).unwrap().meters();
+                let pref_distance = self.config.geolocation_minmax_distance_km as f64 * 1000.0;
+
+                // Map distance to some levels of rating - now they are taken arbitrarily but
+                // they should be somehow related to the distance.
+                if self.config.use_closer_geolocation {
+                    match distance {
+                        _ if distance < pref_distance => rating = NORMALIZE_TO_VALUE,
+                        _ if distance < 2.0 * pref_distance => {
+                            rating = NORMALIZE_TO_VALUE * 2.0 / 3.0
+                        }
+                        _ if distance < 3.0 * pref_distance => {
+                            rating = NORMALIZE_TO_VALUE * 1.0 / 3.0
+                        }
+                        _ => rating = 0.0,
+                    }
+                } else {
+                    match distance {
+                        _ if distance < 0.5 * pref_distance => rating = 0.0,
+                        _ if distance < pref_distance => rating = NORMALIZE_TO_VALUE / 2.0,
+                        _ => rating = NORMALIZE_TO_VALUE,
+                    }
+                }
+            }
+            ratings[node_idx].2 += rating * self.config.location_weight;
+        }
+    }
+
+    fn construct_graph(&self, nodes: &[Node], agraph: &AGraph) -> Graph<IpAddr> {
+        let mut graph = Graph::new();
+
+        for i in 0..agraph.len() {
+            for j in 0..agraph[i].len() {
+                let edge = Edge::new(
+                    IpAddr::from_str(nodes[i].ip.as_str()).unwrap(),
+                    IpAddr::from_str(nodes[j].ip.as_str()).unwrap(),
+                );
+                graph.insert(edge);
+            }
+        }
+        graph
+    }
+
+    fn degree_centrality_avg(&self, degrees: &HashMap<IpAddr, u32>) -> f64 {
+        let sum = degrees
+            .iter()
+            .fold(0.0, |acc, (_, &degree)| acc + degree as f64);
```

Comment:
this fits in one line
```suggestion
        let sum = degrees.iter().fold(0, |acc, (_, &degree)| acc + degree) as f64;
```

Reply:
Done